In [1]:
import numpy as np
import pandas as pd

from scipy.stats import mode

import warnings 
warnings.filterwarnings("ignore")

In [2]:
! pip install pandas


[notice] A new release of pip is available: 24.0 -> 24.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
DATASET_PATH = './housing.csv'
PREPARED_DATASET_PATH = './housing_prepared.csv'

In [4]:
df = pd.read_csv(DATASET_PATH, sep=',')
df.head(10)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   
5    -122.25     37.85                52.0        919.0           213.0   
6    -122.25     37.84                52.0       2535.0           489.0   
7    -122.25     37.84                52.0       3104.0           687.0   
8    -122.26     37.84                42.0       2555.0           665.0   
9    -122.25     37.84                52.0       3549.0           707.0   

   population  households  median_income  median_house_value ocean_proximity  \
0       322.0       126.0         8.3252            452600.0        NEAR BAY   
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY   
2       496.0       177.0         7.2574            352100.0        NEAR BAY   
3       558.0       219.0         5.6431            341300.0        NEAR BAY   
4       565.0       259.0         3.8462            342200.0        NEAR BAY   
5       413.0       193.0         4.0368            269700.0        NEAR BAY   
6      1094.0       514.0         3.6591            299200.0        NEAR BAY   
7      1157.0       647.0         3.1200            241400.0        NEAR BAY   
8      1206.0       595.0         2.0804            226700.0        NEAR BAY   
9      1551.0       714.0         3.6912            261100.0        NEAR BAY   

   id  
0   0  
1   1  
2   2  
3   3  
4   4  
5   5  
6   6  
7   7  
8   8  
9   9

In [5]:
df.tail(10)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
20630    -121.32     39.29                11.0       2640.0           505.0   
20631    -121.40     39.33                15.0       2655.0           493.0   
20632    -121.45     39.26                15.0       2319.0           416.0   
20633    -121.53     39.19                27.0       2080.0           412.0   
20634    -121.56     39.27                28.0       2332.0           395.0   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
20630      1257.0       445.0         3.5673            112000.0   
20631      1200.0       432.0         3.5179            107200.0   
20632      1047.0       385.0         3.1250            115600.0   
20633         NaN       382.0         2.5495             98300.0   
20634      1041.0       344.0         3.7125            116800.0   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637         NaN       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity     id  
20630          INLAND  20630  
20631          INLAND  20631  
20632          INLAND  20632  
20633          INLAND  20633  
20634          INLAND  20634  
20635          INLAND  20635  
20636          INLAND  20636  
20637          INLAND  20637  
20638          INLAND  20638  
20639          INLAND  20639

In [6]:
df.sample(5) # случацные обьекты

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
16724    -120.66     35.47                18.0       2474.0           449.0   
11051    -117.86     33.79                31.0       3523.0           922.0   
7107     -118.01     33.89                34.0       1653.0           292.0   
9330     -122.50     37.96                16.0        100.0            20.0   
12625    -121.54     38.49                 6.0       9104.0          1535.0   

       population  households  median_income  median_house_value  \
16724      1269.0       431.0         3.9063            184800.0   
11051      2660.0       949.0         3.1792            146400.0   
7107       1003.0       310.0         4.6000            203400.0   
9330         45.0        25.0         6.1359            212500.0   
12625      3759.0      1481.0         5.1442            174500.0   

      ocean_proximity     id  
16724       <1H OCEAN  16724  
11051       <1H OCEAN  11051  
7107        <1H OCEAN   7107  
9330         NEAR BAY   9330  
12625          INLAND  12625

In [7]:
df.sample(frac = 0.5) # frac - доля дата сета 1 = 100%, 0,5 = 50% и тд.

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8891     -118.50     34.02                24.0       2924.0          1013.0   
16050    -122.48     37.76                48.0       2304.0           558.0   
6155     -117.95     34.08                34.0       2278.0           476.0   
5256     -118.48     34.07                37.0       4042.0           549.0   
6536     -118.06     34.04                28.0       1516.0           363.0   
...          ...       ...                 ...          ...             ...   
19877    -119.25     36.36                16.0       3245.0           469.0   
6287     -117.90     34.04                15.0      11989.0          2185.0   
8906     -118.48     34.03                39.0       1530.0           401.0   
7445     -118.16     33.94                32.0       2210.0           456.0   
9643     -120.85     37.06                31.0       2609.0           645.0   

       population  households  median_income  median_house_value  \
8891       1492.0       943.0         2.7750            291700.0   
16050      1273.0       512.0         3.2750            332100.0   
6155       1728.0       448.0         3.1250            154100.0   
5256       1318.0       542.0        12.8665            500001.0   
6536       1011.0       344.0         2.6288            160300.0   
...           ...         ...            ...                 ...   
19877      1471.0       450.0         5.8673            154800.0   
6287       6652.0      2081.0         4.5554            278300.0   
8906       1074.0       375.0         3.5076            381800.0   
7445       1270.0       484.0         4.7708            178600.0   
9643       1796.0       629.0         1.5479             82000.0   

      ocean_proximity     id  
8891        <1H OCEAN   8891  
16050        NEAR BAY  16050  
6155        <1H OCEAN   6155  
5256        <1H OCEAN   5256  
6536        <1H OCEAN   6536  
...               ...    ...  
19877          INLAND  19877  
6287        <1H OCEAN   6287  
8906        <1H OCEAN   8906  
7445        <1H OCEAN   7445  
9643           INLAND   9643  

[10320 rows x 11 columns]

In [8]:
df.shape

(20640, 11)

In [9]:
df.columns # просмотр признаков 

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value', 'ocean_proximity', 'id'],
      dtype='object')

In [10]:
df.index

RangeIndex(start=0, stop=20640, step=1)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  19918 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20041 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
 10  id                  20640 non-null  int64  
dtypes: float64(9), int64(1), object(1)
memory usage: 1.7+ MB


In [12]:
df.isna().sum() # пропуски

longitude               0
latitude                0
housing_median_age    722
total_rooms             0
total_bedrooms        207
population            599
households              0
median_income           0
median_house_value      0
ocean_proximity         0
id                      0
dtype: int64

In [13]:
# обращаемся к признаку 
df['households']

0         126.0
1        1138.0
2         177.0
3         219.0
4         259.0
          ...  
20635     330.0
20636     114.0
20637     433.0
20638     349.0
20639     530.0
Name: households, Length: 20640, dtype: float64

In [14]:
df.households # при пробее или точки в имени не рабатает 

0         126.0
1        1138.0
2         177.0
3         219.0
4         259.0
          ...  
20635     330.0
20636     114.0
20637     433.0
20638     349.0
20639     530.0
Name: households, Length: 20640, dtype: float64

In [15]:
df[['households', 'population']]

households  population
0           126.0       322.0
1          1138.0      2401.0
2           177.0       496.0
3           219.0       558.0
4           259.0       565.0
...           ...         ...
20635       330.0       845.0
20636       114.0       356.0
20637       433.0         NaN
20638       349.0       741.0
20639       530.0      1387.0

[20640 rows x 2 columns]

фильтрация с помощью булевых масок

In [16]:
df['population'] > 400

0        False
1         True
2         True
3         True
4         True
         ...  
20635     True
20636    False
20637    False
20638     True
20639     True
Name: population, Length: 20640, dtype: bool

In [17]:
df[df['population']>400].sample(10) # sample перемешивает

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
3739     -118.39     34.18                42.0       1957.0           389.0   
6751     -118.08     34.10                32.0       2830.0           645.0   
15689    -122.43     37.79                52.0       6186.0          1566.0   
12805    -121.46     38.61                43.0       1111.0           269.0   
4295     -118.29     34.09                29.0       2240.0           792.0   
6434     -118.01     34.11                32.0       1978.0           536.0   
7955     -118.16     33.88                18.0       2287.0           662.0   
6560     -118.13     34.19                48.0       2539.0           425.0   
3221     -119.66     36.33                10.0       1623.0           409.0   
19284    -122.73     38.37                40.0       1389.0           309.0   

       population  households  median_income  median_house_value  \
3739        985.0       414.0         2.9327            240200.0   
6751       1500.0       527.0         3.0819            214600.0   
15689      2065.0      1374.0         5.8543            500001.0   
12805       613.0       290.0         1.2917             66300.0   
4295       2254.0       739.0         2.3317            172500.0   
6434        826.0       470.0         2.5114            212200.0   
7955       1804.0       537.0         1.9903            170300.0   
6560        930.0       364.0         4.7269            303900.0   
3221        988.0       395.0         1.4194             58100.0   
19284       841.0       288.0         3.1094            183300.0   

      ocean_proximity     id  
3739        <1H OCEAN   3739  
6751        <1H OCEAN   6751  
15689        NEAR BAY  15689  
12805          INLAND  12805  
4295        <1H OCEAN   4295  
6434           INLAND   6434  
7955        <1H OCEAN   7955  
6560           INLAND   6560  
3221           INLAND   3221  
19284       <1H OCEAN  19284

In [18]:
df[(df['population']>100) & (df['households']>100)].head(10)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   
5    -122.25     37.85                52.0        919.0           213.0   
6    -122.25     37.84                52.0       2535.0           489.0   
7    -122.25     37.84                52.0       3104.0           687.0   
8    -122.26     37.84                42.0       2555.0           665.0   
9    -122.25     37.84                52.0       3549.0           707.0   

   population  households  median_income  median_house_value ocean_proximity  \
0       322.0       126.0         8.3252            452600.0        NEAR BAY   
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY   
2       496.0       177.0         7.2574            352100.0        NEAR BAY   
3       558.0       219.0         5.6431            341300.0        NEAR BAY   
4       565.0       259.0         3.8462            342200.0        NEAR BAY   
5       413.0       193.0         4.0368            269700.0        NEAR BAY   
6      1094.0       514.0         3.6591            299200.0        NEAR BAY   
7      1157.0       647.0         3.1200            241400.0        NEAR BAY   
8      1206.0       595.0         2.0804            226700.0        NEAR BAY   
9      1551.0       714.0         3.6912            261100.0        NEAR BAY   

   id  
0   0  
1   1  
2   2  
3   3  
4   4  
5   5  
6   6  
7   7  
8   8  
9   9

In [19]:
df[(df['population'] > 100) | (df['households'] >100)].head(10)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   
5    -122.25     37.85                52.0        919.0           213.0   
6    -122.25     37.84                52.0       2535.0           489.0   
7    -122.25     37.84                52.0       3104.0           687.0   
8    -122.26     37.84                42.0       2555.0           665.0   
9    -122.25     37.84                52.0       3549.0           707.0   

   population  households  median_income  median_house_value ocean_proximity  \
0       322.0       126.0         8.3252            452600.0        NEAR BAY   
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY   
2       496.0       177.0         7.2574            352100.0        NEAR BAY   
3       558.0       219.0         5.6431            341300.0        NEAR BAY   
4       565.0       259.0         3.8462            342200.0        NEAR BAY   
5       413.0       193.0         4.0368            269700.0        NEAR BAY   
6      1094.0       514.0         3.6591            299200.0        NEAR BAY   
7      1157.0       647.0         3.1200            241400.0        NEAR BAY   
8      1206.0       595.0         2.0804            226700.0        NEAR BAY   
9      1551.0       714.0         3.6912            261100.0        NEAR BAY   

   id  
0   0  
1   1  
2   2  
3   3  
4   4  
5   5  
6   6  
7   7  
8   8  
9   9

In [20]:
~(df['population']>100) # тильда нот все наоборот

0        False
1        False
2        False
3        False
4        False
         ...  
20635    False
20636    False
20637     True
20638    False
20639    False
Name: population, Length: 20640, dtype: bool

In [21]:
(df['population'] > 100)

0         True
1         True
2         True
3         True
4         True
         ...  
20635     True
20636     True
20637    False
20638     True
20639     True
Name: population, Length: 20640, dtype: bool

In [22]:
df[~(df['population'] > 100)].head() 

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
34     -122.27     37.83                51.0       2665.0           574.0   
59     -122.29     37.82                 2.0        158.0            43.0   
61     -122.29     37.82                49.0        135.0            29.0   
73     -122.29     37.81                46.0         12.0             4.0   
121    -122.24     37.85                37.0        334.0            54.0   

     population  households  median_income  median_house_value  \
34          NaN       536.0         2.7303            109700.0   
59         94.0        57.0         2.5625             60000.0   
61         86.0        23.0         6.1183             75000.0   
73         18.0         7.0         0.4999             67500.0   
121        98.0        47.0         4.9643            335000.0   

    ocean_proximity   id  
34         NEAR BAY   34  
59         NEAR BAY   59  
61         NEAR BAY   61  
73         NEAR BAY   73  
121        NEAR BAY  121

In [23]:
# сохранить отдельную выборку в отдельный дата фрейм
tmp = df[~(df['population'] > 100)]
tmp.head()



longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
34     -122.27     37.83                51.0       2665.0           574.0   
59     -122.29     37.82                 2.0        158.0            43.0   
61     -122.29     37.82                49.0        135.0            29.0   
73     -122.29     37.81                46.0         12.0             4.0   
121    -122.24     37.85                37.0        334.0            54.0   

     population  households  median_income  median_house_value  \
34          NaN       536.0         2.7303            109700.0   
59         94.0        57.0         2.5625             60000.0   
61         86.0        23.0         6.1183             75000.0   
73         18.0         7.0         0.4999             67500.0   
121        98.0        47.0         4.9643            335000.0   

    ocean_proximity   id  
34         NEAR BAY   34  
59         NEAR BAY   59  
61         NEAR BAY   61  
73         NEAR BAY   73  
121        NEAR BAY  121

In [24]:
# смотрим индесы созданного дата фрейма
tmp.index

Index([   34,    59,    61,    73,   121,   129,   140,   167,   194,   201,
       ...
       20452, 20453, 20484, 20489, 20502, 20571, 20624, 20625, 20633, 20637],
      dtype='int64', length=828)

In [25]:
# loc  по индексу обьекта старого датафрейма, может быть строкой.
# iloc - принимает порядковый индекс. первый обьект индек ноль и так по порядку

tmp.loc[34]

longitude              -122.27
latitude                 37.83
housing_median_age        51.0
total_rooms             2665.0
total_bedrooms           574.0
population                 NaN
households               536.0
median_income           2.7303
median_house_value    109700.0
ocean_proximity       NEAR BAY
id                          34
Name: 34, dtype: object

In [26]:
tmp.iloc[0]

longitude              -122.27
latitude                 37.83
housing_median_age        51.0
total_rooms             2665.0
total_bedrooms           574.0
population                 NaN
households               536.0
median_income           2.7303
median_house_value    109700.0
ocean_proximity       NEAR BAY
id                          34
Name: 34, dtype: object

In [27]:
#срез с помощью loc
tmp.loc[:61] #по индексам старого датафейма

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
34    -122.27     37.83                51.0       2665.0           574.0   
59    -122.29     37.82                 2.0        158.0            43.0   
61    -122.29     37.82                49.0        135.0            29.0   

    population  households  median_income  median_house_value ocean_proximity  \
34         NaN       536.0         2.7303            109700.0        NEAR BAY   
59        94.0        57.0         2.5625             60000.0        NEAR BAY   
61        86.0        23.0         6.1183             75000.0        NEAR BAY   

    id  
34  34  
59  59  
61  61

In [28]:
tmp.iloc[:4] # по порядковому номеру

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
34    -122.27     37.83                51.0       2665.0           574.0   
59    -122.29     37.82                 2.0        158.0            43.0   
61    -122.29     37.82                49.0        135.0            29.0   
73    -122.29     37.81                46.0         12.0             4.0   

    population  households  median_income  median_house_value ocean_proximity  \
34         NaN       536.0         2.7303            109700.0        NEAR BAY   
59        94.0        57.0         2.5625             60000.0        NEAR BAY   
61        86.0        23.0         6.1183             75000.0        NEAR BAY   
73        18.0         7.0         0.4999             67500.0        NEAR BAY   

    id  
34  34  
59  59  
61  61  
73  73

In [29]:
# индексация может быть отрицательная. 
# возьмем последние пять обьектов

tmp.iloc[-6 : -1]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
20502    -118.68     34.33                45.0        121.0            25.0   
20571    -121.99     38.53                 6.0       4598.0           834.0   
20624    -121.41     39.04                16.0       1698.0           300.0   
20625    -121.52     39.12                37.0        102.0            17.0   
20633    -121.53     39.19                27.0       2080.0           412.0   

       population  households  median_income  median_house_value  \
20502        67.0        27.0         2.9821            325000.0   
20571         NaN       812.0         3.4186            127300.0   
20624         NaN       291.0         3.0739             87200.0   
20625        29.0        14.0         4.1250             72000.0   
20633         NaN       382.0         2.5495             98300.0   

      ocean_proximity     id  
20502       <1H OCEAN  20502  
20571          INLAND  20571  
20624          INLAND  20624  
20625          INLAND  20625  
20633          INLAND  20633

pivot_tables - создание сводных таблиц

In [30]:
# в данном примере высчитываем
# среднее по функции среднего мат.
# по количеству 'total_rooms'
# на пересечении параметров 'ocean_proximity' столбец
# и 'housing_median_age' стает строкой

pd.pivot_table(data = df,  values = 'total_rooms',
                           index = 'ocean_proximity',
                           columns='housing_median_age',
                           aggfunc = 'mean'
                           )

housing_median_age     1.0          2.0          3.0          4.0   \
ocean_proximity                                                      
-                       NaN          NaN          NaN          NaN   
<1H OCEAN               NaN  6401.789474  4422.000000  6364.053333   
INLAND              1101.25  5609.533333  7826.806452  6194.164384   
ISLAND                  NaN          NaN          NaN          NaN   
NEAR BAY                NaN   712.000000  6623.000000  4549.750000   
NEAR OCEAN              NaN  3626.500000  6032.000000  6647.840000   

housing_median_age         5.0          6.0          7.0          8.0   \
ocean_proximity                                                          
-                           NaN          NaN  1659.000000          NaN   
<1H OCEAN           5290.195122  4895.355932  6394.565217  4433.000000   
INLAND              5110.294118  4365.310811  4938.484848  3915.522936   
ISLAND                      NaN          NaN          NaN          NaN   
NEAR BAY            5140.533333  3219.285714  9565.500000  4166.571429   
NEAR OCEAN          3082.526316  4707.200000  5294.470588  3073.555556   

housing_median_age         9.0          10.0  ...         43.0         44.0  \
ocean_proximity                               ...                             
-                           NaN          NaN  ...          NaN  1934.000000   
<1H OCEAN           3518.205882  3539.333333  ...  1664.389222  1682.994413   
INLAND              3979.943925  3545.041096  ...  1515.840000  1756.453125   
ISLAND                      NaN          NaN  ...          NaN          NaN   
NEAR BAY            3413.750000  3447.473684  ...  2038.586957  1991.456522   
NEAR OCEAN          4097.826087  3425.500000  ...  1892.844828  1815.714286   

housing_median_age         45.0         46.0         47.0         48.0  \
ocean_proximity                                                          
-                           NaN          NaN          NaN          NaN   
<1H OCEAN           1676.606897  1829.878505  1716.526882  1811.365854   
INLAND              1564.139535  1616.830189  1767.794118  1618.945946   
ISLAND                      NaN          NaN          NaN          NaN   
NEAR BAY            2086.377358  1879.571429  1885.585366  1839.694444   
NEAR OCEAN          1863.288889  1723.250000  1760.583333  1630.117647   

housing_median_age         49.0         50.0         51.0         52.0  
ocean_proximity                                                         
-                   1141.000000          NaN          NaN          NaN  
<1H OCEAN           1852.320755  1838.890909  1755.384615  1694.745283  
INLAND              2045.533333  1627.818182  1891.916667  1509.414201  
ISLAND                      NaN          NaN          NaN  1827.333333  
NEAR BAY            1796.976744  1851.621622  1923.733333  2082.142623  
NEAR OCEAN          1932.941176  1879.263158  1942.833333  1643.606299  

[6 rows x 52 columns]

In [31]:
df[
    (df.housing_median_age == 3)
    &
    (df.ocean_proximity == 'ISLAND')
]['total_rooms'].count()

0

Приведение типоданных

In [32]:
df.dtypes

longitude             float64
latitude              float64
housing_median_age    float64
total_rooms           float64
total_bedrooms        float64
population            float64
households            float64
median_income         float64
median_house_value    float64
ocean_proximity        object
id                      int64
dtype: object

In [33]:
type(df.longitude) #просматриваем инфу о данных сипиес или дф имеются разные функции
# для разной структуры данных

pandas.core.series.Series

In [34]:
# смотрим тип данных отдельного признака
df['id'].dtype

dtype('int64')

In [35]:
# смотрим из каких значений состоит признак
df['id'].values

array([    0,     1,     2, ..., 20637, 20638, 20639], dtype=int64)

In [36]:
type(df['id'].values)

numpy.ndarray

## модель работает по числовым значениям. если мы оставим числовые значения то модель подумает что это важное значение и построит свою логику на ложно важных значениях. чтобы этого не допустить переводим наименее важные в строку

In [37]:
df['id'] = df['id'].astype(str)
df['id'].dtype

dtype('O')

количественные переменные или количественные признаки

количественные (непременные) перемееные - это те переменные которые могут принять любые числовые значения в определенном диапозоне.

имет типы:
# количествееные
(измеряем  числа)
* дискретные -целочисленные значения (количество людей и т.д) 
* непрерывные - то что можем выразить в вещественном значении ( время, рост и т. д)

# качественные
(строки. порядковые)
* номинальные - то что можем ранжировать видя ирерхическую взаимозависимость
* порядковая - можем ранжировать и задать ирархичность



с количественными и качественными работаем по разному

In [38]:
# создаем датасет только с определенным типом данных
df_num_features = df.select_dtypes(include=['float64', 'int64'])
df_num_features.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value  
0       322.0       126.0         8.3252            452600.0  
1      2401.0      1138.0         8.3014            358500.0  
2       496.0       177.0         7.2574            352100.0  
3       558.0       219.0         5.6431            341300.0  
4       565.0       259.0         3.8462            342200.0

In [39]:
# с помощью .describe смотрим что из себя
# представляют данные в датасете
df_num_features.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        19918.000000  20640.000000   
mean    -119.471242     35.036934           28.653630   2635.763081   
std        5.041408     94.903955           12.576796   2181.615252   
min     -124.350000 -13534.030000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max      122.030000   1327.130000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20041.000000  20640.000000   20640.000000   
mean       537.870553   1425.418243    499.539680       3.870671   
std        421.385070   1135.185798    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    786.000000    280.000000       2.563400   
50%        435.000000   1165.000000    409.000000       3.534800   
75%        647.000000   1726.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

In [40]:
#смотрим для отдельного признака
df['total_rooms'].describe()

count    20640.000000
mean      2635.763081
std       2181.615252
min          2.000000
25%       1447.750000
50%       2127.000000
75%       3148.000000
max      39320.000000
Name: total_rooms, dtype: float64

In [41]:
# смотрим квантиль
np.quantile(df['total_rooms'], q = 0.5)

2127.0

In [42]:
# медиана
df['total_rooms'].median()

2127.0

In [43]:
np.quantile(df['total_rooms'], q=0.25)

1447.75

In [44]:
df['total_rooms'].mode()[0]

1527.0

In [45]:
df['total_rooms'].value_counts()

total_rooms
1527.0     18
1613.0     17
1582.0     17
2127.0     16
1717.0     15
           ..
9614.0      1
10839.0     1
11872.0     1
6205.0      1
10035.0     1
Name: count, Length: 5926, dtype: int64

In [46]:
df['ocean_proximity'].value_counts()

ocean_proximity
<1H OCEAN     9127
INLAND        6542
NEAR OCEAN    2655
NEAR BAY      2288
-               23
ISLAND           5
Name: count, dtype: int64

In [47]:
df[df['total_rooms']== 1527].shape

(18, 11)

Обзор номинативных(котегориальных)- тип дискретной переменной. качественная(неколичественная) переменная величина

In [48]:
#также создаем датафрейм с определенным
# типом данных на этот раз object
df_obji_features = df.select_dtypes(include='object')
df_obji_features.head()

ocean_proximity id
0        NEAR BAY  0
1        NEAR BAY  1
2        NEAR BAY  2
3        NEAR BAY  3
4        NEAR BAY  4

In [49]:
df['ocean_proximity'].value_counts()

ocean_proximity
<1H OCEAN     9127
INLAND        6542
NEAR OCEAN    2655
NEAR BAY      2288
-               23
ISLAND           5
Name: count, dtype: int64

In [50]:
# смотрим уникальные значения
df['ocean_proximity'].unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', '-', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [51]:
# или получаем их количество
df['ocean_proximity'].nunique()

6

Заполнение пустых значений

In [52]:
# смотрим пропуски
df.isna().sum()

longitude               0
latitude                0
housing_median_age    722
total_rooms             0
total_bedrooms        207
population            599
households              0
median_income           0
median_house_value      0
ocean_proximity         0
id                      0
dtype: int64

самым незаполненным является housing_median_age

создадим вспомогательный признак и заполним его нулями если пропуска в housing_median_age нет. и единицами если пропуск имеется

In [53]:
# создаем фиктивный признак. 
# модель поймет что значения заменено
df['housing_median_age_nan'] = 0 # изначально признак заполняем нулями
df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637         NaN       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity     id  housing_median_age_nan  
0            NEAR BAY      0                       0  
1            NEAR BAY      1                       0  
2            NEAR BAY      2                       0  
3            NEAR BAY      3                       0  
4            NEAR BAY      4                       0  
...               ...    ...                     ...  
20635          INLAND  20635                       0  
20636          INLAND  20636                       0  
20637          INLAND  20637                       0  
20638          INLAND  20638                       0  
20639          INLAND  20639                       0  

[20640 rows x 12 columns]

In [54]:
# isna присвоит значение True где есть пропуски
# .loc - срез
df.loc[df['housing_median_age'].isna(), 'housing_median_age_nan']

25       0
117      0
138      0
170      0
299      0
        ..
20531    0
20536    0
20600    0
20601    0
20604    0
Name: housing_median_age_nan, Length: 722, dtype: int64

In [55]:
# и меняем на еденицу
df.loc[df['housing_median_age'].isna(), 'housing_median_age_nan'] = 1

In [56]:
# проверяем выборку
df[df['housing_median_age_nan'] ==1]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
25       -122.28     37.85                 NaN        535.0           123.0   
117      -122.25     37.83                 NaN       4190.0          1105.0   
138      -122.20     37.82                 NaN       1229.0           181.0   
170      -122.25     37.79                 NaN        629.0           188.0   
299      -122.17     37.76                 NaN       1764.0           397.0   
...          ...       ...                 ...          ...             ...   
20531    -121.81     38.58                 NaN       1964.0           314.0   
20536    -121.73     38.55                 NaN       1717.0           393.0   
20600    -121.57     39.16                 NaN       1632.0           367.0   
20601    -121.57     39.13                 NaN        442.0           103.0   
20604    -121.54     39.12                 NaN       4251.0           899.0   

       population  households  median_income  median_house_value  \
25          317.0       119.0         2.4038            107500.0   
117        1786.0      1037.0         3.0897            234100.0   
138         420.0       176.0         7.0175            366700.0   
170         742.0       196.0         2.6458            125000.0   
299         987.0       354.0         2.4333             98200.0   
...           ...         ...            ...                 ...   
20531       808.0       286.0         5.9629            286000.0   
20536      1224.0       387.0         2.7917            130800.0   
20600       769.0       330.0         3.1029             71700.0   
20601       413.0        88.0         1.5694             57900.0   
20604      3265.0       934.0         2.3496             65000.0   

      ocean_proximity     id  housing_median_age_nan  
25           NEAR BAY     25                       1  
117          NEAR BAY    117                       1  
138          NEAR BAY    138                       1  
170          NEAR BAY    170                       1  
299          NEAR BAY    299                       1  
...               ...    ...                     ...  
20531          INLAND  20531                       1  
20536          INLAND  20536                       1  
20600          INLAND  20600                       1  
20601          INLAND  20601                       1  
20604          INLAND  20604                       1  

[722 rows x 12 columns]

In [57]:
median = df['housing_median_age'].median()
median

29.0

In [58]:
df['housing_median_age'].fillna(median)

0        41.0
1        21.0
2        52.0
3        52.0
4        52.0
         ... 
20635    25.0
20636    18.0
20637    17.0
20638    18.0
20639    16.0
Name: housing_median_age, Length: 20640, dtype: float64

In [59]:
df['housing_median_age'].isna().sum()

722

In [60]:
df['housing_median_age'] = df['housing_median_age'].fillna(median)

In [61]:
df.isna().sum()

longitude                   0
latitude                    0
housing_median_age          0
total_rooms                 0
total_bedrooms            207
population                599
households                  0
median_income               0
median_house_value          0
ocean_proximity             0
id                          0
housing_median_age_nan      0
dtype: int64

total_bedrooms 

In [62]:
median = df['total_bedrooms'].median()
df['total_bedrooms'].fillna(median, inplace=True)

In [63]:
df['total_bedrooms'].isna().sum()

0

делаем все вместе

In [64]:
medians = df[['total_bedrooms', 'housing_median_age', 'population']].median()
medians

total_bedrooms         435.0
housing_median_age      29.0
population            1165.0
dtype: float64

In [65]:
df[['total_bedrooms', 'housing_median_age', 'population']] = \
    df[['total_bedrooms', 'housing_median_age', 'population']].fillna(medians)

In [66]:
df.isna().sum()

longitude                 0
latitude                  0
housing_median_age        0
total_rooms               0
total_bedrooms            0
population                0
households                0
median_income             0
median_house_value        0
ocean_proximity           0
id                        0
housing_median_age_nan    0
dtype: int64

до этого работали с количественными признаками, теперь с котегориальными

ocean_proximity

In [67]:
#вновь создаем фиктивный признак с нулями
df['ocean_proximity_nan'] = 0
df.loc[df['ocean_proximity'] =='-', 'ocean_proximity_nan'] = 1

In [68]:
df[df['ocean_proximity_nan'] == 1].head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
1153    -121.46     39.54                14.0       5549.0          1000.0   
2435    -119.59     36.57                19.0       1733.0           303.0   
2636    -124.15     40.59                39.0       1186.0           238.0   
5980    -117.74     34.10                26.0       2723.0           604.0   
6373    -118.02     34.15                44.0       2419.0           437.0   

      population  households  median_income  median_house_value  \
1153      1822.0       919.0         2.9562            142300.0   
2435       911.0       281.0         3.5987            131700.0   
2636       539.0       212.0         2.0938             79600.0   
5980      1847.0       498.0         2.6779            136000.0   
6373      1045.0       432.0         3.8750            280800.0   

     ocean_proximity    id  housing_median_age_nan  ocean_proximity_nan  
1153               -  1153                       0                    1  
2435               -  2435                       0                    1  
2636               -  2636                       0                    1  
5980               -  5980                       0                    1  
6373               -  6373                       0                    1

заменяем модой самым наиболее часто встречающимся числом

In [69]:
# просмотрим моду
df['ocean_proximity'].mode()

0    <1H OCEAN
Name: ocean_proximity, dtype: object

проверяем убеждаемся

In [70]:
df['ocean_proximity'].value_counts()

ocean_proximity
<1H OCEAN     9127
INLAND        6542
NEAR OCEAN    2655
NEAR BAY      2288
-               23
ISLAND           5
Name: count, dtype: int64

In [71]:
df['ocean_proximity'].mode()[0]

'<1H OCEAN'

In [72]:
# заменяем '-' в признаке 'ocean_proximity' на моду
df.replace(
    {'ocean_proximity':
     {'-' : df['ocean_proximity'].mode()[0]}
    },
    inplace = True)

In [73]:
df['ocean_proximity'].value_counts()

ocean_proximity
<1H OCEAN     9150
INLAND        6542
NEAR OCEAN    2655
NEAR BAY      2288
ISLAND           5
Name: count, dtype: int64

Обработка выбросов

In [74]:
df.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.471242     35.036934           28.665746   2635.763081   
std        5.041408     94.903955           12.355019   2181.615252   
min     -124.350000 -13534.030000            1.000000      2.000000   
25%     -121.800000     33.930000           19.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max      122.030000   1327.130000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20640.000000  20640.000000  20640.000000   20640.000000   
mean       536.838857   1417.860562    499.539680       3.870671   
std        419.391878   1119.445348    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        297.000000    797.000000    280.000000       2.563400   
50%        435.000000   1165.000000    409.000000       3.534800   
75%        643.250000   1701.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  housing_median_age_nan  ocean_proximity_nan  
count        20640.000000            20640.000000         20640.000000  
mean        206855.816909                0.034981             0.001114  
std         115395.615874                0.183735             0.033364  
min          14999.000000                0.000000             0.000000  
25%         119600.000000                0.000000             0.000000  
50%         179700.000000                0.000000             0.000000  
75%         264725.000000                0.000000             0.000000  
max         500001.000000                1.000000             1.000000

широта(latitude) земли принимает значения от −90° до 90°. 0° – широта экватора; −90° – широта Южного полюса; 90° – широта Северного полюса
долгота(longitude)-принимает значения от −180° до 180°. Положительные значения соответствуют восточной долготе (сокращённо в.д. или E); отрицательные – западной долготе (сокращённо з.д. или W)._

координаты калифорнии гуглим получаем 
* широта(latitude) 35,1258000
* долгота(longitude) -117,9859000

In [75]:
# longitude - долгота калифорнии отрицательная
# все что + не верно, смотрим
df[df['longitude'] >= 0]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
3479      118.51     34.29                29.0       1287.0           194.0   
5904      118.43     34.29                39.0       1769.0           410.0   
8405      118.36     33.93                40.0       1625.0           500.0   
8636      118.41     33.88                43.0       2492.0           449.0   
13051     121.29     38.61                17.0      13553.0          2474.0   
15263     117.27     33.02                21.0       2144.0           340.0   
17085       0.00     37.47                33.0       1266.0           415.0   
17359       0.00     34.88                 4.0       3680.0           559.0   
18551     122.03     36.96                28.0       1607.0           421.0   
19423       0.00     37.69                 5.0       9601.0          1639.0   

       population  households  median_income  median_house_value  \
3479        525.0       187.0         6.4171            319300.0   
5904       1499.0       390.0         3.1212            153500.0   
8405       2036.0       476.0         2.6298            156500.0   
8636       1033.0       437.0         7.9614            500001.0   
13051      6544.0      2359.0         3.9727            132700.0   
15263       928.0       344.0         5.7980            286100.0   
17085      1991.0       334.0         2.9200            202800.0   
17359      1678.0       569.0         5.0639            201700.0   
18551       926.0       385.0         2.4250            216100.0   
19423      4449.0      1575.0         4.5332            195500.0   

      ocean_proximity     id  housing_median_age_nan  ocean_proximity_nan  
3479        <1H OCEAN   3479                       0                    0  
5904        <1H OCEAN   5904                       0                    0  
8405        <1H OCEAN   8405                       0                    0  
8636        <1H OCEAN   8636                       0                    0  
13051          INLAND  13051                       0                    0  
15263      NEAR OCEAN  15263                       0                    0  
17085      NEAR OCEAN  17085                       0                    0  
17359       <1H OCEAN  17359                       0                    0  
18551      NEAR OCEAN  18551                       0                    0  
19423          INLAND  19423                       0                    0

In [76]:
# создаем фиктивный признак заполненый нулями,
# делаем маску в которой положительные значения 
# долготы заменяем на еденицу  
df['longitude_outlier'] = 0
df.loc[df['longitude'] >= 0, 'longitude_outlier'] = 1

In [77]:
df[df['longitude'] >=0].head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
3479      118.51     34.29                29.0       1287.0           194.0   
5904      118.43     34.29                39.0       1769.0           410.0   
8405      118.36     33.93                40.0       1625.0           500.0   
8636      118.41     33.88                43.0       2492.0           449.0   
13051     121.29     38.61                17.0      13553.0          2474.0   

       population  households  median_income  median_house_value  \
3479        525.0       187.0         6.4171            319300.0   
5904       1499.0       390.0         3.1212            153500.0   
8405       2036.0       476.0         2.6298            156500.0   
8636       1033.0       437.0         7.9614            500001.0   
13051      6544.0      2359.0         3.9727            132700.0   

      ocean_proximity     id  housing_median_age_nan  ocean_proximity_nan  \
3479        <1H OCEAN   3479                       0                    0   
5904        <1H OCEAN   5904                       0                    0   
8405        <1H OCEAN   8405                       0                    0   
8636        <1H OCEAN   8636                       0                    0   
13051          INLAND  13051                       0                    0   

       longitude_outlier  
3479                   1  
5904                   1  
8405                   1  
8636                   1  
13051                  1

In [78]:
#ошибка сборщика данных потерял минус т.к 
# положительные значения попадают
# под отрицательные кординаты калифорнии
df.loc[df['longitude'] > 0, 'longitude'] * -1

3479    -118.51
5904    -118.43
8405    -118.36
8636    -118.41
13051   -121.29
15263   -117.27
18551   -122.03
Name: longitude, dtype: float64

In [79]:
#добавляем значения в исходный дата фрейм
df.loc[df['longitude'] > 0,
       'longitude'] = df.loc[df['longitude'] > 0, 'longitude'] * -1

In [80]:
#все нулевые значения кординат заменяем на медиану
df.loc[df['longitude'] == 0] = df['longitude'].median()

In [81]:
#  с помощю маски проверяем
df[df['longitude']>=0]

Empty DataFrame
Columns: [longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, median_house_value, ocean_proximity, id, housing_median_age_nan, ocean_proximity_nan, longitude_outlier]
Index: []

In [82]:
# latitude калифорнии = 35,1258000 создам маску
df[(df['latitude'] <= 0) | (df['latitude'] > 50)]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8283     -118.13 -13534.03               45.00      1016.00          172.00   
12772    -121.42   1327.13               29.00      2217.00          536.00   
17085    -118.49   -118.49             -118.49      -118.49         -118.49   
17359    -118.49   -118.49             -118.49      -118.49         -118.49   
19423    -118.49   -118.49             -118.49      -118.49         -118.49   

       population  households  median_income  median_house_value  \
8283       361.00      163.00         7.5000           434500.00   
12772     1203.00      507.00         1.9412            73100.00   
17085     -118.49     -118.49      -118.4900             -118.49   
17359     -118.49     -118.49      -118.4900             -118.49   
19423     -118.49     -118.49      -118.4900             -118.49   

      ocean_proximity      id  housing_median_age_nan  ocean_proximity_nan  \
8283       NEAR OCEAN    8283                    0.00                 0.00   
12772          INLAND   12772                    1.00                 0.00   
17085         -118.49 -118.49                 -118.49              -118.49   
17359         -118.49 -118.49                 -118.49              -118.49   
19423         -118.49 -118.49                 -118.49              -118.49   

       longitude_outlier  
8283                0.00  
12772               0.00  
17085            -118.49  
17359            -118.49  
19423            -118.49

In [83]:
# снова создаем фиктивный признак
#  и заполняем единицами
df['latitude_outlier'] = 0
df.loc[(df['latitude'] <= 0) | (df['latitude'] > 50), 'latitude_outlier'] = 1

In [85]:
df[(df['latitude']<= 0) | (df['latitude'] > 50)]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8283     -118.13 -13534.03               45.00      1016.00          172.00   
12772    -121.42   1327.13               29.00      2217.00          536.00   
17085    -118.49   -118.49             -118.49      -118.49         -118.49   
17359    -118.49   -118.49             -118.49      -118.49         -118.49   
19423    -118.49   -118.49             -118.49      -118.49         -118.49   

       population  households  median_income  median_house_value  \
8283       361.00      163.00         7.5000           434500.00   
12772     1203.00      507.00         1.9412            73100.00   
17085     -118.49     -118.49      -118.4900             -118.49   
17359     -118.49     -118.49      -118.4900             -118.49   
19423     -118.49     -118.49      -118.4900             -118.49   

      ocean_proximity      id  housing_median_age_nan  ocean_proximity_nan  \
8283       NEAR OCEAN    8283                    0.00                 0.00   
12772          INLAND   12772                    1.00                 0.00   
17085         -118.49 -118.49                 -118.49              -118.49   
17359         -118.49 -118.49                 -118.49              -118.49   
19423         -118.49 -118.49                 -118.49              -118.49   

       longitude_outlier  latitude_outlier  
8283                0.00                 1  
12772               0.00                 1  
17085            -118.49                 1  
17359            -118.49                 1  
19423            -118.49                 1

In [86]:
df.loc[(df['latitude']<= 0) | (df['latitude'] > 50), 'latitude'] = df['latitude'].median()

In [87]:
df[(df['latitude']<= 0) | (df['latitude'] > 50)]

Empty DataFrame
Columns: [longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, median_house_value, ocean_proximity, id, housing_median_age_nan, ocean_proximity_nan, longitude_outlier, latitude_outlier]
Index: []

total_rooms

Найдем выбросы с помощью квантилей. Наблюдения, значения которых больше 97,5% квантиля и которые меньше 2,5% квантиля будем считать выбросами и заменять на медиану.

In [88]:
df['total_rooms'].describe()

count    20640.000000
mean      2635.041062
std       2181.296146
min       -118.490000
25%       1447.000000
50%       2126.500000
75%       3146.250000
max      39320.000000
Name: total_rooms, dtype: float64

In [89]:
df['total_rooms'].value_counts()

total_rooms
1527.0     18
1613.0     17
1582.0     17
2127.0     16
1717.0     15
           ..
15622.0     1
9614.0      1
10839.0     1
11872.0     1
10035.0     1
Name: count, Length: 5926, dtype: int64

In [90]:
total_rooms_max_value = np.quantile(df['total_rooms'], q = 0.975)
total_rooms_max_value

8068.024999999998

In [91]:
total_rooms_min_value = np.quantile(df['total_rooms'], q = 0.025)
total_rooms_min_value

369.975

In [93]:
condition = (df['total_rooms'] > total_rooms_max_value) | (df['total_rooms'] < total_rooms_min_value)
df[condition]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
59       -122.29     37.82                 2.0        158.0            43.0   
61       -122.29     37.82                49.0        135.0            29.0   
73       -122.29     37.81                46.0         12.0             4.0   
81       -122.27     37.81                52.0        210.0            56.0   
82       -122.28     37.81                52.0        340.0            97.0   
...          ...       ...                 ...          ...             ...   
20578    -121.81     38.84                37.0        352.0            65.0   
20620    -121.48     39.05                40.0        198.0            41.0   
20625    -121.52     39.12                37.0        102.0            17.0   
20627    -121.32     39.13                 5.0        358.0            65.0   
20629    -121.39     39.12                28.0      10035.0          1856.0   

       population  households  median_income  median_house_value  \
59           94.0        57.0         2.5625             60000.0   
61           86.0        23.0         6.1183             75000.0   
73           18.0         7.0         0.4999             67500.0   
81          183.0        56.0         1.1667            112500.0   
82          200.0        87.0         1.5208            112500.0   
...           ...         ...            ...                 ...   
20578       238.0        67.0         2.8542            275000.0   
20620       151.0        48.0         4.5625            100000.0   
20625        29.0        14.0         4.1250             72000.0   
20627       169.0        59.0         3.0000            162500.0   
20629      6912.0      1818.0         2.0943            108300.0   

      ocean_proximity     id  housing_median_age_nan  ocean_proximity_nan  \
59           NEAR BAY     59                     0.0                  0.0   
61           NEAR BAY     61                     0.0                  0.0   
73           NEAR BAY     73                     0.0                  0.0   
81           NEAR BAY     81                     0.0                  0.0   
82           NEAR BAY     82                     0.0                  0.0   
...               ...    ...                     ...                  ...   
20578          INLAND  20578                     0.0                  0.0   
20620          INLAND  20620                     0.0                  0.0   
20625          INLAND  20625                     0.0                  0.0   
20627          INLAND  20627                     0.0                  0.0   
20629          INLAND  20629                     0.0                  0.0   

       longitude_outlier  latitude_outlier  
59                   0.0                 0  
61                   0.0                 0  
73                   0.0                 0  
81                   0.0                 0  
82                   0.0                 0  
...                  ...               ...  
20578                0.0                 0  
20620                0.0                 0  
20625                0.0                 0  
20627                0.0                 0  
20629                0.0                 0  

[1032 rows x 15 columns]

видим что у 1032 обьекта есть аномалии по признаку total_rooms

In [95]:
df.loc[condition, 'total_rooms'] = df['total_rooms'].median()
df['total_rooms'].describe()

count    20640.000000
mean      2435.097045
std       1357.848395
min        370.000000
25%       1511.000000
50%       2126.500000
75%       3005.000000
max       8068.000000
Name: total_rooms, dtype: float64

отбор и построение новых признаков

In [96]:
df.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value', 'ocean_proximity', 'id', 'housing_median_age_nan',
       'ocean_proximity_nan', 'longitude_outlier', 'latitude_outlier'],
      dtype='object')

In [97]:
# удаляем id
df.drop(columns = 'id', inplace  = True)    

Количественные переменные

In [98]:
# доля спален в общем количестве комнат
df['bedroom_share'] = df['total_bedrooms'] / df['total_rooms'] * 100
# сколько человек в среднем живут в одной комнате
df['population_per_room'] = df['population'] / df['total_rooms']

In [99]:
df[['bedroom_share', 'population_per_room']].head()

bedroom_share  population_per_room
0      14.659091             0.365909
1      15.579659             0.338217
2      12.951602             0.338105
3      18.445840             0.437991
4      17.209588             0.347265